sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/EnumTrailingCommaCheck.java

```diff
@@ -0,0 +1,162 @@
+ * valid according to the Java Spec (see <a
+ * href="http://docs.oracle.com/javase/specs/jls/se8/html/jls-8.html#jls-8.9.1">Enum Constants</a>)
+ *
+ * <pre>
+ * enum Type {
+ *     ALPHA,
+ *     BETA,
+ *     GAMMA,
+ *     DELTA, // removing this comma will result in a violation with the check activated
+ * }
+ * </pre>
+ *
+ * <p>However, you could also add a semicolon behind that comma on the same line, which would raise
+ * a violation
+ *
+ * <pre>
+ * enum Type {
+ *     ALPHA,
+ *     BETA,
+ *     GAMMA,
+ *     DELTA,; // violation
+ * }
+ * </pre>
+ * In this case the semicolon should be removed. However, if there is more in the enum body, the
+ * semicolon should be placed on a line by itself.
+ *
+ * <p>An example of how to configure the check is:
+ * <pre>
+ * &lt;module name="EnumTrailingComma"/&gt;
+ * </pre>
+ *
+ * <p>Please note that using this check together with {@code NoWhitespaceBefore} or
+ * {@code SeparatorWrap} may create conflicts with enums that contain a body:
+ * {@code EnumTrailingComma} enforces the semicolon on a separate line while
+ * {@code NoWhiteSpaceBefore} does not allow the semicolon to be preceded with whitespace and
+ * {@code SeparatorWrap} expects the semicolon to be on the same line as the last enum constant.
+ *
+ * @author <a href="kariem.hussein@gmail.com">Kariem Hussein</a>
+ */
+public class EnumTrailingCommaCheck extends AbstractCheck {
+
+    /** Key for warning message text in "messages.properties" file. */
+    public static final String MSG_KEY = "enum.trailing.comma";
+
+    @Override
+    public int[] getDefaultTokens() {
+        return getAcceptableTokens();
+    }
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return new int[] {TokenTypes.ENUM_DEF};
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return getAcceptableTokens();
+    }
+
+    @Override
+    public void visitToken(DetailAST enumDef) {
+        final DetailAST enumConstBlock = enumDef.findFirstToken(TokenTypes.OBJBLOCK);
+
+        final DetailAST enumConstLeft = enumConstBlock.findFirstToken(TokenTypes.LCURLY);
+        final DetailAST enumConstRight = enumConstBlock.findFirstToken(TokenTypes.RCURLY);
+
+        // Only check, if block is multi-line and there are more than one enum constants
+        if (enumConstLeft.getLineNo() != enumConstRight.getLineNo()
+                && enumConstBlock.getChildCount(TokenTypes.ENUM_CONSTANT_DEF) > 1) {
+            final DetailAST constant = enumConstBlock.findFirstToken(TokenTypes.ENUM_CONSTANT_DEF);
+            final DetailAST lastComma = getLastComma(constant);
+
+            final DetailAST nextAst = lastComma.getNextSibling();
+            if (isIllegalTokenAfterComma(lastComma, nextAst)) {
+                log(nextAst, MSG_KEY);
+            }
+        }
+    }
+
+    private boolean isIllegalTokenAfterComma(DetailAST lastComma, DetailAST nextAst) {
```

Comment:
please explain why this method is not `isSmthAfterCommaOnSameLine` with no special processing of SEMI.
I think it should be checking for any sibling on same line.

Reply:
According to the Java Language Specification on [Enum Constants](https://docs.oracle.com/javase/specs/jls/se8/html/jls-8.html#jls-8.9.1) the only allowed symbols are

1. Other enum constants, separated by a comma
2. Enum body declarations, starting with a semicolon

If this method does not allow any sibling on the same line, this would also include traditional comments (starting with `TokenType.BLOCK_COMMENT_BEGIN`) and end-of-line comments (`TokenType.SINGLE_LINE_COMMENT`). Both, however, should be allowed. Am I misunderstanding something?

Reply:
@romani Just checked the implementation again. Good to know that we do not need to check for comments, so the explicit check on symbols seems to be redundant.

However, after looking at this again, I still think the current version is very clear on what is done: we have two problem cases:

1. The last comma is followed by an enum constant: this means that there is no more comma after that enum constant, because we have already checked for the last comma

2. The last comma is followed by a semicolon on the same line.

This is the implementation

``` java
private boolean isIllegalTokenAfterComma(DetailAST lastComma, DetailAST nextAst) {
    boolean illegalToken = false;

    final int nextAstType = nextAst.getType();
    if (nextAstType == TokenTypes.SEMI) {
        if (nextAst.getLineNo() == lastComma.getLineNo()) {
            // semi on the same line as last comma
            illegalToken = true;
        }
    }
    else if (nextAstType == TokenTypes.ENUM_CONSTANT_DEF) {
        // next enum constant on the same line after comma
        illegalToken = true;
    }
    return illegalToken;
}
```

I have now tried different options and could not find a clearer solution, so I do not really understand what you are asking for when you do not want the "extra validatoin" of tokens. We will always have to check for the semicolon or the right curly bracket to see whether we are at the end of the enum constant block.

Also, this method could be a lot less verbose so I have just updated it. See be8be03107a778e9d91c70341a29fc7c5c21a356

``` java
private boolean isIllegalTokenAfterComma(DetailAST lastComma, DetailAST nextAst) {
    final int nextAstType = nextAst.getType();

    // semi on the same line as last comma, or followed by enum constant
    return (nextAstType == TokenTypes.SEMI && nextAst.getLineNo() == lastComma.getLineNo())
            || nextAstType == TokenTypes.ENUM_CONSTANT_DEF;
}
```

Does that work for you?

Reply:
I have rewritten that whole block. It's now similar to the ArrayTrailingCommaCheck, with some additional comments:

`enumConstEnd` can either be SEMI or RCURLY.

``` java
// Only check, if block is multi-line and there are more than one enum constants
if (enumConstLeft.getLineNo() != enumConstEnd.getLineNo()
        && enumConstBlock.getChildCount(TokenTypes.ENUM_CONSTANT_DEF) > 1) {

    final DetailAST lastSibling = enumConstEnd.getPreviousSibling();
    if (lastSibling.getType() != TokenTypes.COMMA) {
        // constant definition does not end in a comma
        log(lastSibling, MSG_KEY);
    } else if (enumConstEnd.getType() == TokenTypes.SEMI
            && enumConstEnd.getLineNo() == lastSibling.getLineNo()) {
        // constant definition ends in a comma, but is followed by semi on same line
        log(enumConstEnd, MSG_KEY);
    }
}
```

With this, we can remove the other helper methods `getLastComma` and `isIllegalTokenAfterComma`.

Does that work for you?

Reply:
@romani I haven't directly answered the question, but I think the current solution is better than I had previously suggested. Let me quickly answer the question

>  Fact that I do not understand is why we care about "last comma"?

We only care about the "last comma". This is the point of the check. Hence the name Enum**TrailingComma**Check. Intermediary commas are required as per language specification. Only the last, the trailing comma, is relevant for this check.